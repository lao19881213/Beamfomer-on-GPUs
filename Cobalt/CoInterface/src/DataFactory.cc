//# DataFactory.cc
//# Copyright (C) 2011-2013  ASTRON (Netherlands Institute for Radio Astronomy)
//# P.O. Box 2, 7990 AA Dwingeloo, The Netherlands
//#
//# This file is part of the LOFAR software suite.
//# The LOFAR software suite is free software: you can redistribute it and/or
//# modify it under the terms of the GNU General Public License as published
//# by the Free Software Foundation, either version 3 of the License, or
//# (at your option) any later version.
//#
//# The LOFAR software suite is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with the LOFAR software suite. If not, see <http://www.gnu.org/licenses/>.
//#
//# $Id: DataFactory.cc 25662 2013-07-13 19:10:59Z mol $

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

#include <CoInterface/DataFactory.h>

#include <CoInterface/OutputTypes.h>
#include <CoInterface/CorrelatedData.h>
#include <CoInterface/BeamFormedData.h>
#include <CoInterface/TriggerData.h>


namespace LOFAR
{
  namespace Cobalt
  {


    StreamableData *newStreamableData(const Parset &parset, OutputType outputType, unsigned streamNr, Allocator &allocator)
    {
      switch (outputType) {
      case CORRELATED_DATA: return new CorrelatedData(parset.nrMergedStations(), parset.nrChannelsPerSubband(), parset.integrationSteps(), allocator, 512); // 512 alignment is needed for writing to MS

      case BEAM_FORMED_DATA: {
        const struct ObservationSettings::BeamFormer::File &file = parset.settings.beamFormer.files[streamNr];
        const struct ObservationSettings::BeamFormer::StokesSettings &sset =
          file.coherent ? parset.settings.beamFormer.coherentSettings
                        : parset.settings.beamFormer.incoherentSettings;

        unsigned nrSubbands = parset.settings.nrSubbands(file.sapNr);
        unsigned nrChannels = sset.nrChannels;
        unsigned nrSamples = parset.settings.nrSamplesPerSubband() / sset.nrChannels / sset.timeIntegrationFactor;

        return new FinalBeamFormedData(nrSamples, nrSubbands, nrChannels, allocator);
      }

      case TRIGGER_DATA: return new TriggerData;

      default: THROW(CoInterfaceException, "unsupported output type");
      }

    }


  } // namespace Cobalt
} // namespace LOFAR

