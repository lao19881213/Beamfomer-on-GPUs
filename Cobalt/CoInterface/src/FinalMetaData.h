//# FinalMetaData.h
//# Copyright (C) 2012-2013  ASTRON (Netherlands Institute for Radio Astronomy)
//# P.O. Box 2, 7990 AA Dwingeloo, The Netherlands
//#
//# This file is part of the LOFAR software suite.
//# The LOFAR software suite is free software: you can redistribute it and/or
//# modify it under the terms of the GNU General Public License as published
//# by the Free Software Foundation, either version 3 of the License, or
//# (at your option) any later version.
//#
//# The LOFAR software suite is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with the LOFAR software suite. If not, see <http://www.gnu.org/licenses/>.
//#
//# $Id: FinalMetaData.h 24388 2013-03-26 11:14:29Z amesfoort $

#ifndef LOFAR_INTERFACE_FINAL_METADATA_H
#define LOFAR_INTERFACE_FINAL_METADATA_H

#include <cstddef>
#include <string>
#include <vector>
#include <ostream>

#include <Stream/Stream.h>

namespace LOFAR
{
  namespace Cobalt
  {

    class FinalMetaData
    {
    public:
      struct BrokenRCU {
        std::string station; // CS001, etc
        std::string type;  // RCU, LBA, HBA
        size_t seqnr;      // RCU/antenna number
        std::string time;  // date time of break

        BrokenRCU()
        {
        }
        BrokenRCU(const std::string &station, const std::string &type, size_t seqnr, const std::string &time) :
          station(station), type(type), seqnr(seqnr), time(time)
        {
        }

        bool operator==(const BrokenRCU &other) const
        {
          return station == other.station && type == other.type && seqnr == other.seqnr && time == other.time;
        }
      };

      std::vector<BrokenRCU>  brokenRCUsAtBegin, brokenRCUsDuring;

      void write(Stream &s);
      void read(Stream &s);
    };

    std::ostream& operator<<(std::ostream& os, const struct FinalMetaData::BrokenRCU &rcu);

    std::ostream& operator<<(std::ostream& os, const FinalMetaData &finalMetaData);

  } // namespace Cobalt
} // namespace LOFAR

#endif

