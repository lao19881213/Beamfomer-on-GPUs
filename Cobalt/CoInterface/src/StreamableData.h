//# StreamableData.h
//# Copyright (C) 2008-2013  ASTRON (Netherlands Institute for Radio Astronomy)
//# P.O. Box 2, 7990 AA Dwingeloo, The Netherlands
//#
//# This file is part of the LOFAR software suite.
//# The LOFAR software suite is free software: you can redistribute it and/or
//# modify it under the terms of the GNU General Public License as published
//# by the Free Software Foundation, either version 3 of the License, or
//# (at your option) any later version.
//#
//# The LOFAR software suite is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with the LOFAR software suite. If not, see <http://www.gnu.org/licenses/>.
//#
//# $Id: StreamableData.h 25445 2013-06-24 06:47:16Z mol $

#ifndef LOFAR_INTERFACE_STREAMABLE_DATA_H
#define LOFAR_INTERFACE_STREAMABLE_DATA_H

#include <cstring>

#include <Common/LofarTypes.h>
#include <Common/DataConvert.h>
#include <CoInterface/Parset.h>
#include <CoInterface/MultiDimArray.h>
#include <CoInterface/SparseSet.h>
#include <CoInterface/Allocator.h>
#include <CoInterface/BlockID.h>
#include <Stream/Stream.h>

namespace LOFAR
{
  namespace Cobalt
  {

    // Data which needs to be transported between CN, ION and Storage.
    // Apart from read() and write() functionality, the data is augmented
    // with a sequence number in order to detect missing data. Furthermore,
    // an integration operator += can be defined to reduce the data.

    // Endianness:
    // * Endianness is defined by the correlator.
    // * Both Data and sequence number will have endianness of the
    //   correlator
    //
    // WARNING: We consider all data streams to have the endianness of the
    // correlator. No conversion is done here.

    class IntegratableData
    {
    public:
      virtual ~IntegratableData()
      {
      }

      virtual IntegratableData &operator += (const IntegratableData &) = 0;
    };


    class StreamableData
    {
    public:
      static const uint32_t magic = 0xda7a;
      static const size_t alignment = 512;

      // Freely modified by GPUProc (only)
      struct BlockID blockID;

      // the CPU which fills the datastructure sets the peerMagicNumber,
      // because other CPUs will overwrite it with a read(s,true) call from
      // either disk or network.
      StreamableData() : peerMagicNumber(magic), rawSequenceNumber(0)
      {
      }
      virtual ~StreamableData()
      {
      }

      void read(Stream *, bool withSequenceNumber, unsigned align = 1);
      void write(Stream *, bool withSequenceNumber, unsigned align = 1);

      bool shouldByteSwap() const
      {
        return peerMagicNumber != magic;
      }

      uint32_t sequenceNumber(bool raw = false) const
      {
        if (shouldByteSwap() && !raw) {
          uint32_t seqno = rawSequenceNumber;

          byteSwap32(&seqno);

          return seqno;
        } else {
          return rawSequenceNumber;
        }
      }

      void setSequenceNumber(uint32_t seqno)
      {
        if (shouldByteSwap())
          byteSwap32(&seqno);

        rawSequenceNumber = seqno;
      }

      virtual void setDimensions(unsigned, unsigned, unsigned)
      {
      }

      uint32_t peerMagicNumber;  /// magic number received from peer

    protected:
      // a subclass should override these to marshall its data
      virtual void readData(Stream *, unsigned) = 0;
      virtual void writeData(Stream *, unsigned) = 0;

    private:
      uint32_t rawSequenceNumber; /// possibly needs byte swapping
    };


    // A typical data set contains a MultiDimArray of tuples and a set of flags.
    template <typename T = fcomplex, unsigned DIM = 4, unsigned FLAGS_DIM = 2>
    class SampleData : public StreamableData
    {
    public:
      typedef typename MultiDimArray<T,DIM>::ExtentList ExtentList;
      typedef typename MultiDimArray<SparseSet<unsigned>,FLAGS_DIM>::ExtentList FlagsExtentList;

      SampleData(const ExtentList &extents, const FlagsExtentList &flagsExtents, Allocator & = heapAllocator);

      MultiDimArray<T,DIM>              samples;
      MultiDimArray<SparseSet<unsigned>,FLAGS_DIM>   flags;

    protected:
      virtual void readData(Stream *, unsigned);
      virtual void writeData(Stream *, unsigned);

    private:
      //bool	 itsHaveWarnedLittleEndian;
    };


    inline void StreamableData::read(Stream *str, bool withSequenceNumber, unsigned alignment)
    {
      if (withSequenceNumber) {
        std::vector<char> header(alignment > 2 * sizeof(uint32_t) ? alignment : 2 * sizeof(uint32_t));
        uint32_t          &magicValue = *reinterpret_cast<uint32_t *>(&header[0]);
        uint32_t          &seqNo = *reinterpret_cast<uint32_t *>(&header[sizeof(uint32_t)]);

        str->read(&header[0], header.size());

        peerMagicNumber = magicValue;
        rawSequenceNumber = seqNo;
      }

      readData(str, alignment);
    }


    inline void StreamableData::write(Stream *str, bool withSequenceNumber, unsigned alignment)
    {

      if (withSequenceNumber) {
        /*     std::vector<char> header(alignment > sizeof(uint32_t) ? alignment : sizeof(uint32_t)); */
        std::vector<char> header(alignment > 2 * sizeof(uint32_t) ? alignment : 2 * sizeof(uint32_t));
        uint32_t          &magicValue = *reinterpret_cast<uint32_t *>(&header[0]);
        uint32_t          &seqNo = *reinterpret_cast<uint32_t *>(&header[sizeof(uint32_t)]);

#if defined USE_VALGRIND
        std::memset(&header[0], 0, header.size());
#endif

        magicValue = peerMagicNumber;
        seqNo = rawSequenceNumber;

        str->write(&header[0], header.size());
      }

      writeData(str, alignment);
    }


    template <typename T, unsigned DIM, unsigned FLAGS_DIM>
    inline SampleData<T,DIM,FLAGS_DIM>::SampleData(const ExtentList &extents, const FlagsExtentList &flagsExtents, Allocator &allocator)
      :
      samples(extents, alignment, allocator),
      flags(flagsExtents) // e.g., for FilteredData [nrChannels][nrStations], sparse dimension [nrSamplesPerIntegration]

      //itsHaveWarnedLittleEndian(false)
    {
    }


    template <typename T, unsigned DIM, unsigned FLAGS_DIM>
    inline void SampleData<T,DIM,FLAGS_DIM>::readData(Stream *str, unsigned alignment)
    {
      (void)alignment;

      str->read(samples.origin(), samples.num_elements() * sizeof(T));
    }


    template <typename T, unsigned DIM, unsigned FLAGS_DIM>
    inline void SampleData<T,DIM,FLAGS_DIM>::writeData(Stream *str, unsigned alignment)
    {
      (void)alignment;

      str->write(samples.origin(), samples.num_elements() * sizeof(T));
    }

  } // namespace Cobalt
} // namespace LOFAR

#endif

