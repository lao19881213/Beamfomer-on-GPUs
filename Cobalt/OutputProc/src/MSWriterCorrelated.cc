//# MSWriterCorrelated.cc: a writer for correlated visibilities
//# Copyright (C) 2012-2013  ASTRON (Netherlands Institute for Radio Astronomy)
//# P.O. Box 2, 7990 AA Dwingeloo, The Netherlands
//#
//# This file is part of the LOFAR software suite.
//# The LOFAR software suite is free software: you can redistribute it and/or
//# modify it under the terms of the GNU General Public License as published
//# by the Free Software Foundation, either version 3 of the License, or
//# (at your option) any later version.
//#
//# The LOFAR software suite is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with the LOFAR software suite. If not, see <http://www.gnu.org/licenses/>.
//#
//# $Id: MSWriterCorrelated.cc 25659 2013-07-13 13:10:10Z mol $

#include <lofar_config.h>

#include "MSWriterCorrelated.h"

#include <sys/types.h>
#include <fcntl.h>
#include <boost/format.hpp>
#include <boost/lexical_cast.hpp>

#include <Common/SystemUtil.h>
#include <MSLofar/FailedTileInfo.h>
#include <CoInterface/CorrelatedData.h>

#include <tables/Tables/Table.h>
#include <casa/Quanta/MVTime.h>

#include "MeasurementSetFormat.h"

using boost::format;
using namespace casa;


namespace LOFAR
{
  namespace Cobalt
  {

    MSWriterCorrelated::MSWriterCorrelated (const std::string &logPrefix, const std::string &msName, const Parset &parset, unsigned subbandIndex, bool isBigEndian)
      :
      MSWriterFile(
        (makeMeasurementSet(logPrefix, msName, parset, subbandIndex, isBigEndian),
         str(format("%s/table.f0data") % msName))),
      itsLogPrefix(logPrefix),
      itsMSname(msName),
      itsParset(parset)
    {
      if (LofarStManVersion > 1) {
        string seqfilename = str(format("%s/table.f0seqnr") % msName);

        try {
          itsSequenceNumbersFile = new FileStream(seqfilename, O_WRONLY | O_CREAT | O_TRUNC, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
        } catch (...) {
          LOG_WARN_STR(itsLogPrefix << "Could not open sequence numbers file " << seqfilename);
        }
      }

#if 0
      // derive baseline names
      std::vector<std::string> stationNames = parset.mergedStationNames();
      std::vector<std::string> baselineNames(parset.nrBaselines());
      unsigned nrStations = stationNames.size();

      // order of baselines as station indices:
      // 0-0, 1-0, 1-1, 2-0, 2-1, 2-2 ... (see RTCP/CNProc/Correlator.cc)

      unsigned bl = 0;

      for(unsigned s1 = 0; s1 < nrStations; s1++)
        for(unsigned s2 = 0; s2 <= s1; s2++)
          //bl = s1 * (s1 + 1) / 2 + stat2 ;
          baselineNames[bl++] = str(format("%s_%s") % stationNames[s1] % stationNames[s2]);
#endif

      itsConfiguration.add("fileFormat",           "AIPS++/CASA");
      itsConfiguration.add("filename",             LOFAR::basename(msName));
      itsConfiguration.add("size",                 "0");
      itsConfiguration.add("location",             parset.getHostName(CORRELATED_DATA, subbandIndex) + ":" + parset.getDirectoryName(CORRELATED_DATA, subbandIndex));

      itsConfiguration.add("percentageWritten",    "0");
      itsConfiguration.add("startTime",            parset.getString("Observation.startTime"));
      itsConfiguration.add("duration",             "0");
      itsConfiguration.add("integrationInterval",  str(format("%f") % parset.IONintegrationTime()));
      itsConfiguration.add("centralFrequency",     str(format("%f") % parset.settings.subbands[subbandIndex].centralFrequency));
      itsConfiguration.add("channelWidth",         str(format("%f") % parset.channelWidth()));
      itsConfiguration.add("channelsPerSubband",   str(format("%u") % parset.nrChannelsPerSubband()));
      itsConfiguration.add("stationSubband",       str(format("%u") % parset.settings.subbands[subbandIndex].stationIdx));
      itsConfiguration.add("subband",              str(format("%u") % subbandIndex));
      itsConfiguration.add("SAP",                  str(format("%u") % parset.settings.subbands[subbandIndex].SAP));
    }


    MSWriterCorrelated::~MSWriterCorrelated()
    {
    }


    void MSWriterCorrelated::makeMeasurementSet(const std::string &logPrefix, const std::string &msName, const Parset &parset, unsigned subbandIndex, bool isBigEndian)
    {
#if defined HAVE_AIPSPP
      MeasurementSetFormat myFormat(parset, 512);

      myFormat.addSubband(msName, subbandIndex, isBigEndian);

      LOG_INFO_STR(logPrefix << "MeasurementSet created");
#endif // defined HAVE_AIPSPP
    }


    void MSWriterCorrelated::write(StreamableData *data)
    {
      CorrelatedData *cdata = dynamic_cast<CorrelatedData*>(data);

      ASSERT( data );
      ASSERT( cdata );

      // Write data
      MSWriterFile::write(data);

      // Write sequence number
      if (itsSequenceNumbersFile != 0) {
        // quick fix: always write to maintain integrity
        unsigned seqnr = data->sequenceNumber(true);

        itsSequenceNumbersFile->write(&seqnr, sizeof seqnr);
      }

      itsNrBlocksWritten++;

      itsConfiguration.replace("size",     str(format("%u") % getDataSize()));
      itsConfiguration.replace("duration", str(format("%f") % ((data->sequenceNumber() + 1) * itsParset.IONintegrationTime())));
      itsConfiguration.replace("percentageWritten", str(format("%u") % percentageWritten()));
    }


    static MVEpoch datetime2epoch(const string &datetime)
    {
      Quantity q;

      if (!MVTime::read(q, datetime))
        return MVEpoch(0);

      return MVEpoch(q);
    }


    void MSWriterCorrelated::augment(const FinalMetaData &finalMetaData)
    {
      ScopedLock sl(MeasurementSetFormat::sharedMutex);

      map<string, FailedTileInfo::VectorFailed> brokenBefore, brokenDuring;

      // fill set of broken hardware at beginning of observation
      for (size_t i = 0; i < finalMetaData.brokenRCUsAtBegin.size(); i++) {
        const struct FinalMetaData::BrokenRCU &rcu = finalMetaData.brokenRCUsAtBegin[i];

        brokenBefore[rcu.station].push_back(FailedTileInfo(rcu.station, rcu.time, datetime2epoch(rcu.time), rcu.type, rcu.seqnr));
      }

      // fill set of hardware that broke during the observation
      for (size_t i = 0; i < finalMetaData.brokenRCUsDuring.size(); i++) {
        const struct FinalMetaData::BrokenRCU &rcu = finalMetaData.brokenRCUsDuring[i];

        brokenDuring[rcu.station].push_back(FailedTileInfo(rcu.station, rcu.time, datetime2epoch(rcu.time), rcu.type, rcu.seqnr));
      }

      LOG_INFO_STR(itsLogPrefix << "Reopening MeasurementSet");

      Table ms(itsMSname, Table::Update);

      vector<FailedTileInfo::VectorFailed> before(FailedTileInfo::antennaConvert(ms, brokenBefore));
      vector<FailedTileInfo::VectorFailed> during(FailedTileInfo::antennaConvert(ms, brokenDuring));

      LOG_INFO_STR(itsLogPrefix << "Writing broken hardware information to MeasurementSet");

      try {
        FailedTileInfo::writeFailed(ms, before, during);
      } catch (Exception &ex) {
        LOG_ERROR_STR("Failed to write broken hardware information: " << ex);
      }
    }


  } // namespace Cobalt
} // namespace LOFAR

