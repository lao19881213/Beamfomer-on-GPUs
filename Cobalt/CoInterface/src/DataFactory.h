//# DataFactory.h
//# Copyright (C) 2011-2013  ASTRON (Netherlands Institute for Radio Astronomy)
//# P.O. Box 2, 7990 AA Dwingeloo, The Netherlands
//#
//# This file is part of the LOFAR software suite.
//# The LOFAR software suite is free software: you can redistribute it and/or
//# modify it under the terms of the GNU General Public License as published
//# by the Free Software Foundation, either version 3 of the License, or
//# (at your option) any later version.
//#
//# The LOFAR software suite is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with the LOFAR software suite. If not, see <http://www.gnu.org/licenses/>.
//#
//# $Id: DataFactory.h 25662 2013-07-13 19:10:59Z mol $

#ifndef LOFAR_INTERFACE_DATA_FACTORY_H
#define LOFAR_INTERFACE_DATA_FACTORY_H

#include <CoInterface/Allocator.h>
#include <CoInterface/OutputTypes.h>
#include <CoInterface/Parset.h>
#include <CoInterface/StreamableData.h>


namespace LOFAR
{
  namespace Cobalt
  {

    StreamableData *newStreamableData(const Parset &, OutputType, unsigned streamNr, Allocator & = heapAllocator);

  } // namespace Cobalt
} // namespace LOFAR

#endif

