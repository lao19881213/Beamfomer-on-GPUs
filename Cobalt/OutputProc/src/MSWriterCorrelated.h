//# MSMriterCorrelated.h: a writer for correlated visibilities
//# Copyright (C) 2012-2013  ASTRON (Netherlands Institute for Radio Astronomy)
//# P.O. Box 2, 7990 AA Dwingeloo, The Netherlands
//#
//# This file is part of the LOFAR software suite.
//# The LOFAR software suite is free software: you can redistribute it and/or
//# modify it under the terms of the GNU General Public License as published
//# by the Free Software Foundation, either version 3 of the License, or
//# (at your option) any later version.
//#
//# The LOFAR software suite is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with the LOFAR software suite. If not, see <http://www.gnu.org/licenses/>.
//#
//# $Id: MSWriterCorrelated.h 24388 2013-03-26 11:14:29Z amesfoort $

#ifndef LOFAR_STORAGE_MSWRITERCORRELATED_H
#define LOFAR_STORAGE_MSWRITERCORRELATED_H

#include <string>
#include <vector>

#include <Stream/FileStream.h>
#include <CoInterface/Parset.h>
#include <CoInterface/SmartPtr.h>
#include <CoInterface/StreamableData.h>

#include "MSWriterFile.h"

namespace LOFAR
{
  namespace Cobalt
  {


    class MSWriterCorrelated : public MSWriterFile
    {
    public:
      MSWriterCorrelated(const std::string &logPrefix, const std::string &msName, const Parset &parset, unsigned subbandIndex, bool isBigEndian);
      ~MSWriterCorrelated();

      virtual void write(StreamableData *data);

      virtual void augment(const FinalMetaData &finalMetaData);

    protected:
      const std::string itsLogPrefix;
      const std::string itsMSname;
      const Parset &itsParset;

      SmartPtr<FileStream>             itsSequenceNumbersFile;

    private:
      void makeMeasurementSet(const std::string &logPrefix, const std::string &msName, const Parset &parset, unsigned subbandIndex, bool isBigEndian);
    };


  }
}

#endif

