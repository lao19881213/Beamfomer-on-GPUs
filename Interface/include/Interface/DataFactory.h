//#  DataFactory.h
//#
//#  Copyright (C) 2007
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: PrintVector.h 16765 2010-11-25 13:27:09Z mol $

#ifndef LOFAR_INTERFACE_DATA_FACTORY_H
#define LOFAR_INTERFACE_DATA_FACTORY_H

#include <Interface/Allocator.h>
#include <Interface/OutputTypes.h>
#include <Interface/Parset.h>
#include <Interface/StreamableData.h>


namespace LOFAR {
namespace RTCP {

StreamableData *newStreamableData(const Parset &, OutputType, int streamNr = -1, Allocator & = heapAllocator);

} // namespace RTCP
} // namespace LOFAR

#endif
