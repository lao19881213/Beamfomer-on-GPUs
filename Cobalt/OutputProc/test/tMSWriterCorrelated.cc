//# tMSWriterCorrelated.cc
//# Copyright (C) 2012-2013  ASTRON (Netherlands Institute for Radio Astronomy)
//# P.O. Box 2, 7990 AA Dwingeloo, The Netherlands
//#
//# This file is part of the LOFAR software suite.
//# The LOFAR software suite is free software: you can redistribute it and/or
//# modify it under the terms of the GNU General Public License as published
//# by the Free Software Foundation, either version 3 of the License, or
//# (at your option) any later version.
//#
//# The LOFAR software suite is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with the LOFAR software suite. If not, see <http://www.gnu.org/licenses/>.
//#
//# $Id: tMSWriterCorrelated.cc 24388 2013-03-26 11:14:29Z amesfoort $

#include <lofar_config.h>

#include <Common/LofarLogger.h>
#include <CoInterface/DataFactory.h>
#include <CoInterface/FinalMetaData.h>
#include <CoInterface/StreamableData.h>
#include <CoInterface/Parset.h>
#include <OutputProc/MSWriterCorrelated.h>

using namespace std;
using namespace LOFAR;
using namespace Cobalt;

#if defined WORDS_BIGENDIAN
const int bigEndian = 1;
#else
const int bigEndian = 0;
#endif

int main()
{
  INIT_LOGGER("tMSWriterCorrelated");

  Parset parset("tMSWriterCorrelated.parset");

  {
    // Create MeasurementSet
    MSWriterCorrelated writer("", "tMSWriterCorrelated.in_1/SB000.MS", parset, 0, bigEndian);

    // Write some data
    StreamableData *data = newStreamableData(parset, CORRELATED_DATA, 0);

    writer.write(data);

    delete data;

    // Add broken tile information
    FinalMetaData fmd;
    struct FinalMetaData::BrokenRCU rcu;

    rcu.station = "CS013";
    rcu.time = "2012-01-01 00:00:00";
    rcu.type = "RCU";
    rcu.seqnr = 2;

    fmd.brokenRCUsAtBegin.push_back(rcu);

    rcu.station = "CS013";
    rcu.time = "2012-01-01 00:00:00";
    rcu.type = "LBA";
    rcu.seqnr = 4;

    fmd.brokenRCUsAtBegin.push_back(rcu);

    writer.augment(fmd);
  }

  return 0;
}

