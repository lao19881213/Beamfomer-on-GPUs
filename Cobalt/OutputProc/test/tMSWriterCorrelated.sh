#!/bin/sh
./runctest.sh tMSWriterCorrelated
