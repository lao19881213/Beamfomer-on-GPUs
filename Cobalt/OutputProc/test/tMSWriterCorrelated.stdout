    select result of 1 rows
1 selected columns:  Col_1
Axis Lengths: [1, 96]  (NB: Matrix in Row/Column order)
[0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]

